lib/src/rules/prefer_if_elements_to_map_literals.dart

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2019, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:linter/src/analyzer.dart';
+
+const _desc = r'Prefer if elements to map literals where possible.';
+
+const _details = r'''
+When building maps from iterables, it is preferable to use `if` and `for` 
+elements rather than conditionals and repetition.
+
+
+**BAD:**
+```
+Map<String, WidgetBuilder>.fromIterable(
+  kAllGalleryDemos,
+  key: (demo) => '${demo.routeName}',
+  value: (demo) => demo.buildRoute,
+);
+
+```
+
+**GOOD:**
+```
+return {
+  for (var demo in kAllGalleryDemos)
+    '${demo.routeName}': demo.buildRoute,
+};
+```
+''';
+
+class PreferIfElementsToMapLiterals extends LintRule implements NodeLintRule {
+  PreferIfElementsToMapLiterals()
+      : super(
+            name: 'prefer_if_elements_to_map_literals',
```

Comment:
Maybe `prefer_if_elements_to_map_fromIterable`?

Reply:
Yes!  Except maybe `prefer_for_elements_to_map_fromIterable` is better still?